app/javascript/pages/UtmLinks/Index.tsx

```diff
@@ -87,71 +97,95 @@ const UtmLinkList = () => {
       setUtmLinksStats((prev) => ({ ...prev, ...stats }));
     })();
   }, 500);
+
   React.useEffect(() => {
     if (utmLinks.length === 0) return;
-    const sortKey = extractSortParam(searchParams)?.key;
+    const sortKey = sort?.key;
     if (sortKey === "sales_count" || sortKey === "revenue_cents" || sortKey === "conversion_rate") return;
     const ids = utmLinks.flatMap((link) =>
       utmLinkWithStats(link, utmLinksStats[link.id]).sales_count === null ? [link.id] : [],
     );
     if (ids.length === 0) return;
 
     debouncedGetUtmLinksStats(ids);
-  }, [utmLinks, searchParams]);
+  }, [utmLinks, sort]);
+
+  const navigateWithParams = (params: { page?: number; query?: string; sort?: Sort<SortKey> | null }) => {
+    const newParams: Record<string, string | number | undefined> = {};
+
+    if (params.page !== undefined && params.page > 1) {
+      newParams.page = params.page;
+    }
+    if (params.query !== undefined && params.query.length > 0) {
+      newParams.query = params.query;
+    }
+    if (params.sort !== undefined && params.sort !== null) {
+      newParams.key = params.sort.key;
+      newParams.direction = params.sort.direction;
+    }
+
+    setIsLoading(true);
+    router.get(
+      Routes.utm_links_dashboard_path(newParams),
+      {},
+      {
+        preserveState: true,
+        preserveScroll: true,
+        onFinish: () => setIsLoading(false),
+      },
+    );
```

Comment:
`router.reload` is what you need here, it preserves scroll and state by default
```suggestion
    router.reload(
      {
        data: newParams,
        onStart: () => setIsLoading(true),
        onFinish: () => setIsLoading(false),
      },
    );
```